// @wordpress/element/build/react-platform.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createPortal", {
  enumerable: true,
  get: function () {
    return _reactDom.createPortal;
  }
});
Object.defineProperty(exports, "createRoot", {
  enumerable: true,
  get: function () {
    return _client.createRoot;
  }
});
Object.defineProperty(exports, "findDOMNode", {
  enumerable: true,
  get: function () {
    return _reactDom.findDOMNode;
  }
});
Object.defineProperty(exports, "flushSync", {
  enumerable: true,
  get: function () {
    return _reactDom.flushSync;
  }
});
Object.defineProperty(exports, "hydrate", {
  enumerable: true,
  get: function () {
    return _reactDom.hydrate;
  }
});
Object.defineProperty(exports, "hydrateRoot", {
  enumerable: true,
  get: function () {
    return _client.hydrateRoot;
  }
});
Object.defineProperty(exports, "render", {
  enumerable: true,
  get: function () {
    return _reactDom.render;
  }
});
Object.defineProperty(exports, "unmountComponentAtNode", {
  enumerable: true,
  get: function () {
    return _reactDom.unmountComponentAtNode;
  }
});
var _reactDom = require("react-dom");
var _client = require("react-dom/client");
